### DISCUSSION TITLE
Provide WebSocket as a transport wrapper of OSC packets

### DISCUSSION THREAD

**Discussion by knah** [vrchat-community/osc#30] [Ideas]
OSC is pretty much a unidirectional UDP-based protocol.

Use of WebSocket to encapsulate OSC messages can provide several benefits:
 * WebSocket already has binary message frames that nicely map 1-to-1 to the same "a bunch of bytes" idea that UDP packets are
 * WebSocket communications are bidirectional, so this solves the issue of possible multiple receivers having to fight for a single UDP port for VRChat-to-external-software communication
 * WebSocket is connection-oriented. VRChat can know if there's a client, and not send any data if there isn't one.
 * In addition to the above, the client can specify the data it's interested in for that connection, so VRChat has to do less work sending all the parameters if the client is interested in only a few
 * In another addition to the above, the client can tell if VRChat is running in the first place, instead of just throwing out UDP packets and hoping for the best
 * WebSocket is usable from browsers. This has both pros and cons:
   * Some external applications can be implemented fully as webpages, which improves end-user security. Web browsers are amazing as far as sandboxing code goes.
   * Some kind of permission/authentication system would be required so that a *random* web page doesn't mess with VRChat unnoticed/unapproved.

One major drawback of a WebSocket implementation is that it uses TCP, and with that comes head-of-line blocking which might affect performance over slow networks. However, within a single system, that is not really a concern.

Naturally, most of this can be provided by a standalone unofficial application that would proxy data, but having official first-party support would be neat. Additionally, first-party support would enable additional features and optimization.

**Comment by momo-the-monster** [comment]
I like websockets and use them in a lot of my own work. [OSCQuery](https://github.com/vrchat-community/osc/milestone/4) will help with some of the discoverability issues. WebSockets present some security issues not present with OSC, as you mention. Nonetheless, this is a reasonable idea and it's under consideration for the long-term.

**Comment by songminwoo1** [comment]
I'm currently considering this. maybe I'll implement this in 2~3 days. I'm currently using web for UI in my VRC OSC multi-port forwarding Config Manager.
https://booth.pm/ko/items/3687929
